lib/client/terminal/terminal_windows.go

```diff
@@ -0,0 +1,279 @@
+//go:build windows
+// +build windows
+
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package terminal
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"syscall"
+
+	"github.com/Azure/go-ansiterm/winterm"
+	"github.com/gravitational/teleport/lib/client/tncon"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/moby/term"
+)
+
+// initTerminal configures the terminal for raw, VT compatible output and
+// optionally input. The returned function should be called before program
+// exit to ensure the terminal is reset, otherwise it will be left in a broken
+// state.
+func initTerminal(input bool) (func(), error) {
+	stdoutFd := int(syscall.Stdout)
+	stdinFd := int(syscall.Stdin)
+
+	oldOutMode, err := winterm.GetConsoleMode(uintptr(stdoutFd))
+	if err != nil {
+		return func() {}, fmt.Errorf("failed to retrieve stdout mode: %w", err)
+	}
+
+	oldInMode, err := winterm.GetConsoleMode(uintptr(stdinFd))
+	if err != nil {
+		return func() {}, fmt.Errorf("failed to retrieve stdout mode: %w", err)
+	}
+
+	newOutMode := oldOutMode | winterm.ENABLE_VIRTUAL_TERMINAL_PROCESSING | winterm.DISABLE_NEWLINE_AUTO_RETURN
+
+	err = winterm.SetConsoleMode(uintptr(stdoutFd), newOutMode)
+	if err != nil {
+		return func() {}, fmt.Errorf("failed to set stdout mode: %w", err)
+	}
+
+	if input {
+		newInMode := oldInMode
+		newInMode &^= winterm.ENABLE_ECHO_INPUT
+		newInMode &^= winterm.ENABLE_LINE_INPUT
+		newInMode &^= winterm.ENABLE_MOUSE_INPUT
+		newInMode &^= winterm.ENABLE_WINDOW_INPUT
+		newInMode &^= winterm.ENABLE_PROCESSED_INPUT
+
+		newInMode |= winterm.ENABLE_EXTENDED_FLAGS
+		newInMode |= winterm.ENABLE_INSERT_MODE
+		newInMode |= winterm.ENABLE_QUICK_EDIT_MODE
+		newInMode |= winterm.ENABLE_VIRTUAL_TERMINAL_INPUT
+
+		err = winterm.SetConsoleMode(uintptr(stdinFd), newInMode)
+		if err != nil {
+			return func() {}, fmt.Errorf("failed to set stdin mode: %w", err)
```

Comment:
In this case, you've set the mode for stdout and are not returning a cleanup func to restore the old mode? Is this what you want?

I can see a client not bothering to call the cleanup func if an error is returned, so maybe you want to just try restoring the stdout mode here before returning the error.

Reply:
Good catch, it now attempts to reset the output mode on error.